src/core/surface/call.c

```diff
@@ -926,9 +932,34 @@ static void finish_batch_completion(grpc_exec_ctx *exec_ctx, void *user_data,
   GRPC_CALL_INTERNAL_UNREF(exec_ctx, call, "completion");
 }
 
+/* Call post_batch_completion() if it was not already called and (if there are
+   no more steps to complete or if the batch failed) */
+static void check_and_call_post_batch_completion(grpc_exec_ctx *exec_ctx,
+                                                 batch_control *bctl) {
+  int are_steps_complete;
+  if (bctl->post_batch_completion_called) {
+    return;
+  }
+
+  /* Note: This decrements bctl->steps_to_complete irrespective of bctl->success
+     value and that is okay) */
+  are_steps_complete = gpr_unref(&bctl->steps_to_complete);
+
+  if (are_steps_complete || bctl->success == 0) {
```

Comment:
Calling post_batch_completion() in case of failure,  even if there are some extra steps left in the batch control feels a bit wrong to me.. (intuitively, it feels like I need to do some more 'cleanup' related to the unfinished steps in the batch)